lib/vnet/daemon/client_darwin.go

```diff
@@ -0,0 +1,197 @@
+	}()
+	C.BundlePath(&result)
+
+	bundlePath := C.GoString(result.bundlePath)
+	if bundlePath == "" {
+		return false, nil
+	}
+
+	if err := exec.CommandContext(ctx, "codesign", "--verify", bundlePath).Run(); err != nil {
+		var exitError *exec.ExitError
+		if errors.As(err, &exitError) {
+			log.DebugContext(ctx, "codesign --verify returned with non-zero exit code",
+				"exit_code", exitError.ExitCode(), "bundle_path", bundlePath)
+
+			if exitError.ExitCode() == codesignExitCodeVerificationFailed {
+				return false, nil
+			}
+
+			// Returning a custom error instead of wrapping err because err is just a cryptic "exit status 2".
+			return false, trace.Errorf("failed to check the signature of tsh, codesign returned with exit code %d", exitError.ExitCode())
+		}
+
+		return false, trace.Wrap(err, "failed to check the signature of tsh")
+	}
+
+	return true, nil
+}
+
+// RegisterAndCall attempts to register the daemon as a login item, waits for the user to enable it
+// and then starts it by sending a message through XPC.
+func RegisterAndCall(ctx context.Context, socketPath, ipv6Prefix, dnsAddr string) error {
+	initialStatus := daemonStatus()
+	// If the status is equal to "requires approval" before RegisterAndCall called register, it means
+	// that it's not the first time the user tries to start the daemon. In that case, macOS is not
+	// going to show the notification about a new login item. Instead, we just open the login items
+	// ourselves to direct the user towards enabling the login item.
+	if initialStatus == serviceStatusRequiresApproval {
+		C.OpenSystemSettingsLoginItems()
+	}
+
+	status, err := register(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if status != serviceStatusEnabled {
+		if err := waitForEnablement(ctx); err != nil {
+			return trace.Wrap(err, "waiting for the login item to get enabled")
+		}
+	}
+
+	return trace.NotImplemented("RegisterAndCall is not fully implemented yet")
+}
+
+func register(ctx context.Context) (serviceStatus, error) {
+	var result C.RegisterDaemonResult
+
+	C.RegisterDaemon(&result)
+	defer func() {
+		C.free(unsafe.Pointer(result.error_description))
+	}()
+
+	if !result.ok {
+		status := serviceStatus(result.service_status)
+		// Docs for registerAndReturnError [1] don't seem to cover a situation in which the user adds
+		// the launch daemon for the first time. In that case, that method returns "Operation not permitted"
+		// error (Code=1, Domain=SMAppServiceErrorDomain). That error doesn't seem to exist in Service
+		// Management Errors [2]. However, all this means is that the launch daemon was added to the
+		// login items and the user must now enable the corresponding login item.
+		//
+		// We can confirm this by looking at the returned service status.
+		//
+		// [1] https://developer.apple.com/documentation/servicemanagement/smappservice/register()?language=objc
+		// [2] https://developer.apple.com/documentation/servicemanagement/service-management-errors?language=objc
```

Comment:
nit: Improper link syntax:

```suggestion
		// [1]: https://developer.apple.com/documentation/servicemanagement/smappservice/register()?language=objc
		// [2]: https://developer.apple.com/documentation/servicemanagement/service-management-errors?language=objc
```